### PR TITLE
Add attention sink support to sparse SDPA

### LINKED ISSUE #54173: Sparse SDPA sink
## Ask

Add an optional `attention_sink` argument to `ttnn.transformer.sparse_sdpa`, with the same semantics
the dense `scaled_dot_product_attention` already has. Needed for DeepSeek-V4 CSA prefill.

## What the sink is

A learned vector, **one scalar per attention head** — nothing per-token, nothing data-dependent.
`[64]` for V4-Flash, `[128]` for V4-Pro
(`models/demos/deepseek_v3_d_p/reference/deepseek_v4/modeling_deepseek_v4.py:797`).

Mathematically it is **one extra term in the softmax denominator**:

```
N[c]  =  Σ_j exp( s[j] ) · v[j, c]        numerator   — unchanged
D     =  Σ_j exp( s[j] )                  denominator

without sink:   out[c]  =  N[c] /  D
with sink:      out[c]  =  N[c] / ( D + exp σ[h] )
```

The reference writes it as a concatenated logit column that gets dropped after softmax
([modeling_deepseek_v4.py](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/deepseek_v3_d_p/reference/deepseek_v4/modeling_deepseek_v4.py):730-743), but that is equivalent to the above

## It is already implemented in the dense kernel

[ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp)

### PR DESCRIPTION
**Pull request by skrsticTT**
Add optional per-head attention sinks to `ttnn.transformer.sparse_sdpa` for DeepSeek-V4 CSA prefill. Matches classic SDPA semantics by including the sink in softmax normalization without contributing to the output numerator.

Closes #54173.

| Pipeline | Status |
| --- | --- |
| Sanity | [Workflow status](https://github.com/tenstorrent/tt-metal/actions/runs/34689630227) |
| L2 nightly · SDPA | [Workflow status](https://github.com/tenstorrent/tt-metal/actions/runs/34689632384) |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:skrstic/sparse-sdpa-attention-sink)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/sparse_sdpa_compute.cpp [hunk 1]
```diff
@@ -498,7 +500,10 @@ void kernel_main() {
                     cb_col_identity,
                     cb_recip_scratch,
                     cb_out_im,
-                    scale_fp32>(sum_cur.get_cb_id(), out_cur.get_cb_id(), Sqt);
+                    scale_fp32,
+                    use_attention_sink,
+                    cb_attention_sink,
+                    /*sink_per_row=*/true>(sum_cur.get_cb_id(), out_cur.get_cb_id(), Sqt, max_cur.get_cb_id());
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/sparse_sdpa_compute.cpp]: sink_per_row, make this a named constexpr variable and pass it in

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/sparse_sdpa_reader.cpp [hunk 2]
```diff
@@ -96,6 +100,37 @@ void kernel_main() {
...
+            scratch.reserve_back(1);
+            noc.async_write_zeros(sink_cb, sink_tiles * tile_bytes);
+            noc.write_zeros_l1_barrier();
+            auto* dst = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(sink_cb.get_write_ptr());
+            auto* src = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(scratch.get_write_ptr());
+            for (uint32_t h = 0; h < H; ++h) {
+                // [1,H,1,1] TILE stores each head in a separate tile. Read the first aligned
+                // segment, then put its scalar in the head's row of a packed first-column tile.
+                noc.async_read(sink, scratch, NOC_DRAM_READ_ALIGNMENT_BYTES, {.page_id = h}, {.offset_bytes = 0});
+                noc.async_read_barrier();
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/sparse_sdpa_reader.cpp]: do we need a barrirer per each read?

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py [hunk 3]
```diff
@@ -560,3 +560,121 @@ def test_sparse_sdpa_bad_layout_rejected_on_hit(device, expect_error):
     tt_q_tile = ttnn.to_layout(tt_q, ttnn.TILE_LAYOUT)  # same shape+dtype (same hash) but wrong layout -> HIT
     with expect_error(RuntimeError, "ROW_MAJOR"):
         ttnn.transformer.sparse_sdpa(tt_q_tile, tt_kv, tt_idx, V_DIM, kv_format=BF16_KV, k_chunk_size=kc)
+
+
+@run_for_blackhole()
+@pytest.mark.parametrize("sink_dtype", [ttnn.bfloat16, ttnn.bfloat8_b, ttnn.bfloat4_b])
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py]: any need for perf/determinism tests?

File: ttnn/cpp/ttnn/operations/transformer/sdpa/sparse_sdpa.hpp [hunk 4]
```diff
@@ -46,6 +46,9 @@ enum class SparseKVFormat : uint8_t {
 //   block_cyclic_cache_tp_sharded : true = the cache is striped across ALL sp*tp devices (linear chip = sp_coord*tp
 //                             + tp_coord), so stripes = sp*tp and per-stripe chunk = chunk_local/tp.
 //
+// attention_sink: optional [1,H,1,1] tiled DRAM BF16/BFP8/BFP4 tensor. Like classic SDPA,
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/sparse_sdpa.hpp]: why is this tiled? seems like it will waste quite some data?

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/sparse_sdpa_device_operation.cpp [hunk 5]
```diff
@@ -27,6 +27,20 @@ void validate_non_hashed(const SparseSDPAParams& attrs, const SparseSDPAInputs&
...
     TT_FATAL(
         q.device() == kv.device() && q.device() == idx.device(), "sparse_sdpa: all inputs must be on the same device");
+    if (t.attention_sink.has_value()) {
+        const auto& sink = t.attention_sink.value();
+        TT_FATAL(
+            sink.storage_type() == StorageType::DEVICE && sink.buffer() != nullptr,
+            "Attention sink tensor must be on device");
+        TT_FATAL(sink.device() == q.device(), "Attention sink must be on the same device as Q");
+        TT_FATAL(
+            sink.layout() == Layout::TILE && sink.dtype() == DataType::BFLOAT16,
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/sparse_sdpa_device_operation.cpp]: nanobind has different info on dtype format?

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/sparse_sdpa_reader.cpp [hunk 6]
```diff
@@ -96,6 +100,37 @@ void kernel_main() {
     const uint32_t idx_l1 = idx_cb.get_write_ptr();
     volatile tt_l1_ptr uint32_t* idx_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(idx_l1);
 
+    if constexpr (use_attention_sink) {
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/sparse_sdpa_reader.cpp]: should this be moved to writer for perf reasons?

File: ttnn/cpp/ttnn/operations/transformer/sdpa/sparse_sdpa.hpp [hunk 7]
```diff
@@ -46,6 +46,9 @@ enum class SparseKVFormat : uint8_t {
...
+// its per-head scalar is multiplied by scale and added only to the softmax denominator.
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/sparse_sdpa.hpp]: DeepSeek-V4 stores the learned sink in the already-scaled attention-logit domain: its reference concatenates `sinks` after `QK * scale`. This contract scales the supplied sink again, matching classic TTNN SDPA, so passing the model weights directly would compute `exp(scale * sink)` instead of `exp(sink)`. Can we explicitly document—and preferably test—that the DeepSeek caller must pass `model_sink / scale`?

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py [hunk 8]
```diff
@@ -560,3 +560,166 @@ def test_sparse_sdpa_bad_layout_rejected_on_hit(device, expect_error):
     tt_q_tile = ttnn.to_layout(tt_q, ttnn.TILE_LAYOUT)  # same shape+dtype (same hash) but wrong layout -> HIT
     with expect_error(RuntimeError, "ROW_MAJOR"):
         ttnn.transformer.sparse_sdpa(tt_q_tile, tt_kv, tt_idx, V_DIM, kv_format=BF16_KV, k_chunk_size=kc)
+
+
+@run_for_blackhole()
```
Comment (ddjekicTT) [inline tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py:565]:
```suggestion
@run_for_blackhole()
@pytest.mark.parametrize(
    "H,S,TOPK,dim,kc",
    [
        pytest.param(128, 4, 128, 64, 32, id="small-boundaries"),
        pytest.param(64, 640, 128 + 512, 512, 128, id="v4-flash-full-selection"),
        pytest.param(128, 640, 128 + 1024, 512, 128, id="v4-pro-full-selection"),
    ],
)
```

Comment (ddjekicTT) [reply tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py:565]: Add these shapes, they are relevant for DS v4 CSA

**Review by ddjekicTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py (hunk 8 above)
Comment (skrsticTT) [reply tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py:565]: any required df?

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py [hunk 9]
```diff
@@ -560,3 +560,166 @@ def test_sparse_sdpa_bad_layout_rejected_on_hit(device, expect_error):
...
+        ttnn.to_torch(tt_q).float(),
+        ttnn.to_torch(tt_kv).float(),
+        indices,
+        scale,
+        dim,
+        attention_sink=ttnn.to_torch(tt_sink).float().reshape(1, H, 1, 1),
+    )
+    actual = ttnn.to_torch(out).float()
+    # Correlation alone would miss the sink's suppression of the output magnitude.
+    torch.testing.assert_close(actual, expected, atol=0.035, rtol=0.08)
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py:588]: The existing sparse-SDPA correctness tests evaluate numerical accuracy with the shared `pcc()` helper, while these new tests use only `torch.testing.assert_close`. Can we add the standard PCC checks to the randomized sink tests?

I agree that PCC alone would not catch an incorrect sink-induced magnitude change, so the targeted `assert_close`/magnitude check should remain as a second assertion. The cache test is an exception because its expected output is constant, where PCC is not meaningful.

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py (hunk 3 above)
Comment (pavlejosipovic) [reply tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py]: bump on this comment

**Review by pavlejosipovic — COMMENTED** [review]
fix clang tidy stuff pls

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py (hunk 8 above)
Comment (ddjekicTT) [reply tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py]: bf16 is good for now

File: tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py (hunk 3 above)
Comment (skrsticTT) [reply tests/ttnn/unit_tests/operations/sdpa/test_sparse_sdpa.py]: yeah i just waited for shapes, will resolve this one when i address it soon

**Review by pavlejosipovic — APPROVED** [review]
